examples/api/lib/material/autocomplete/autocomplete.5.dart

```diff
@@ -0,0 +1,249 @@
+    _currentQuery = query;
+
+    // In a real application, there should be some error handling here.
+    final Iterable<String> options = await _FakeAPI.search(_currentQuery!);
+
+    // If another search happened after this one, throw away these options.
+    if (_currentQuery != query) {
+      return null;
+    }
+    _currentQuery = null;
+
+    return options;
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _debouncedSearch = _debounce<Iterable<String>?, String>(_search);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Autocomplete<String>(
+      showOptionsViewOnUncompletedOptions: true,
+      showOptionsViewOnEmptyOptions: true,
+      optionsBuilder: (TextEditingValue textEditingValue) async {
+        _isFirstFocus = textEditingValue.text.isEmpty;
+        _nothingFound = false;
+        if (_isFirstFocus) {
+          return const Iterable<String>.empty();
+        }
+        _lastValue = textEditingValue.text;
+        _isLoading = true; // Begin options loading.
+
+        final Iterable<String>? options =
+            await _debouncedSearch(textEditingValue.text);
+
+        if (textEditingValue.text == _lastValue) {
+          _isLoading = false; // End options loading.
+        }
+        if (options == null) {
+          return _lastOptions;
+        }
+        _lastOptions = options;
+        _nothingFound = options.isEmpty;
+        return options;
+      },
+      onSelected: (String selection) {
+        debugPrint('You just selected $selection');
+      },
+      optionsViewBuilder: (BuildContext context,
+          AutocompleteOnSelected<String> onSelected, Iterable<String> options) {
+        return _isLoading
+            ? const Text('loading...')
+            : _isFirstFocus
+                ? const Text('Type something')
+                : _nothingFound
+                    ? const Text('no options found!')
+                    : AutocompleteOverlay(
+                        onSelected: onSelected, options: options);
+      },
+    );
+  }
+}
+
+// Mimics a remote API.
+class _FakeAPI {
+  static const List<String> _kOptions = <String>[
+    'aardvark',
+    'bobcat',
+    'chameleon',
+  ];
+
+  // Searches the options, but injects a fake "network" delay.
+  static Future<Iterable<String>> search(String query) async {
+    await Future<void>.delayed(fakeAPIDuration); // Fake 1 second delay.
+    if (query == '') {
+      return const Iterable<String>.empty();
+    }
+    return _kOptions.where((String option) {
+      return option.contains(query.toLowerCase());
+    });
+  }
+}
+
+typedef _Debounceable<S, T> = Future<S?> Function(T parameter);
+
+/// Returns a new function that is a debounced version of the given function.
+///
+/// This means that the original function will be called only after no calls
+/// have been made for the given Duration.
+_Debounceable<S, T> _debounce<S, T>(_Debounceable<S?, T> function) {
```

Comment:
If you totally removed the debouncing from this example, would it be easier to understand the new stuff, or would it be too unrealistic?

Reply:
Yes, but it might also make it difficult to understand why `_isLoading` should be set to false only when the current text is equal to the last text searched with (`_lastValue`). So, if someone were to dup the example, remove the (seemingly) extra variables `_lastOptions` and `_lastValue` and later add debouncing, they would probably run into errors and get confused.